### PR TITLE
1) add dotenv, 2) add timing logic for concurrent and sequential exec…

### PR DESCRIPTION
**Pull request by Tata0703**
…utions, 3) refactor code

### CODE REVIEW

**Comment by Tata0703** [comment]
Does agent_execution_time mean agent_turnaround_time? @dongyuanjushi

**Comment by Tata0703** [comment]
For `python eval.py --llm_name gemini-pro --agents MathAgent:5,NarrativeAgent:3,RecAgent:2` on Apple M3 Pro
```
Concurrent Metrics: {'agent_turnaround_time': {'avg': 93.96543960571289, 'p90': 116.99870007038116, 'p99': 126.38843154191972}, 'agent_waiting_time': {'avg': 8.624463605880738, 'p90': 17.499514484405516, 'p99': 18.15974139213562}, 'request_waiting_time': {'avg': 21.310650926828384, 'p90': 31.25480697154999, 'p99': 42.31232394456863}, 'request_turnaround_time': {'avg': 24.77129928469658, 'p90': 32.746853375434874, 'p99': 53.824659507274625}}
Sequential Metrics: {'agent_turnaround_time': {'avg': 700.6242356777191, 'p90': 1929.82242155075, 'p99': 3349.3627397537234}, 'agent_waiting_time': {'avg': 688.4109686851501, 'p90': 1920.0403087139123, 'p99': 3341.242487788201}, 'request_waiting_time': {'avg': 316.8659665723642, 'p90': 960.0200780868543, 'p99': 1749.5768236446381}, 'request_turnaround_time': {'avg': 320.0195253392061, 'p90': 964.1815799713147, 'p99': 1752.9462733268738}}
Improvement of agent_turnaround_time_avg_improv: 86.59%
Improvement of agent_turnaround_time_p90_improv: 93.94%
Improvement of agent_turnaround_time_p99_improv: 96.23%
Improvement of agent_waiting_time_avg_improv: 98.75%
Improvement of agent_waiting_time_p90_improv: 99.09%
Improvement of agent_waiting_time_p99_improv: 99.46%
Improvement of request_waiting_time_avg_improv: 93.27%
Improvement of request_waiting_time_p90_improv: 96.74%
Improvement of request_waiting_time_p99_improv: 97.58%
Improvement of request_turnaround_time_avg_improv: 92.26%
Improvement of request_turnaround_time_p90_improv: 96.60%
Improvement of request_turnaround_time_p99_improv: 96.93%
```

File: src/utils/calculator.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+            agent_turnaround_time = output["agent_execution_time"] + accumulated_time
+            agent_waiting_time = output["agent_waiting_time"] + accumulated_time
+            rounds = output["rounds"]
+            fixed_offset = accumulated_time / rounds
+
+            # Adjust times by the accumulated time
+            request_waiting_times = [time + fixed_offset for time in output["request_waiting_times"]]
+            request_turnaround_times = [time + fixed_offset for time in output["request_turnaround_times"]]
+
```
Comment (Tata0703) [inline src/utils/calculator.py:79]: @dongyuanjushi could you help check the logic here?

**Comment by Tata0703** [comment]
For `python eval.py --llm_name gpt-4 --agents MathAgent:1,NarrativeAgent:1,RecAgent:1` on Apple M3 Pro

```
Concurrent Metrics: {'agent_turnaround_time': {'avg': 91.69690124193828, 'p90': 102.15579552650452, 'p99': 105.85800618648528}, 'agent_waiting_time': {'avg': 7.9563623269399, 'p90': 17.10281376838684, 'p99': 20.203812770843506}, 'request_waiting_time': {'avg': 15.22387844324112, 'p90': 23.261526823043823, 'p99': 28.86870121240616}, 'request_turnaround_time': {'avg': 25.072641173998516, 'p90': 33.74934582710266, 'p99': 34.976864211559295}}
**** concurrent Execution Statistics Ends ****

**** sequential Execution Statistics Starts ****
Sequential Metrics: {'agent_turnaround_time': {'avg': 70.59157721201579, 'p90': 110.25403418540955, 'p99': 117.32871781826019}, 'agent_waiting_time': {'avg': 36.17042191823324, 'p90': 77.89839305877686, 'p99': 92.08435758590699}, 'request_waiting_time': {'avg': 17.260201546880936, 'p90': 46.8303759098053, 'p99': 46.83044869184494}, 'request_turnaround_time': {'avg': 26.921158366733128, 'p90': 59.98320825099945, 'p99': 60.437594175338745}}
**** sequential Execution Statistics Ends ****

**** Improvement Analysis Starts ****
Improvement of agent_turnaround_time_avg_improv: -29.90%
Improvement of agent_turnaround_time_p90_improv: 7.35%
Improvement of agent_turnaround_time_p99_improv: 9.78%
Improvement of agent_waiting_time_avg_improv: 78.00%
Improvement of agent_waiting_time_p90_improv: 78.04%
Improvement of agent_waiting_time_p99_improv: 78.06%
Improvement of request_waiting_time_avg_improv: 11.80%
Improvement of request_waiting_time_p90_improv: 50.33%
Improvement of request_waiting_time_p99_improv: 38.35%
Improvement of request_turnaround_time_avg_improv: 6.87%
Improvement of request_turnaround_time_p90_improv: 43.74%
Improvement of request_turnaround_time_p99_improv: 42.13%
```

File: src/utils/calculator.py (hunk 1 above)
Comment (dongyuanjushi) [reply src/utils/calculator.py:79]: 1. accumulated time should be: accumulated_time += (agent_turnaround_time - agent_waiting_time)
2.  request_waiting_times and request_turnaround_times should be that only the element at the first position in request level waiting times and turnaround times should add the accumulated_time and the remaining element should not be changed. As the agent will be executed sequential after the whole execution of its previous agent, so only the first request will be much longer, the remaining request waiting time and turnaround time should be the same?

**Review by dongyuanjushi — APPROVED** [review]
accumulated time should be: accumulated_time += (agent_turnaround_time - agent_waiting_time)
request_waiting_times and request_turnaround_times should be that only the element at the first position in request level waiting times and turnaround times should add the accumulated_time and the remaining element should not be changed. As the agent will be executed sequential after the whole execution of its previous agent, so only the first request will be much longer, the remaining request waiting time and turnaround time should be the same. 
I will fix this after fixing the returned agent_turnaround_time in OpenAGI first.

**Pull request merged**